### DISCUSSION TITLE
Changing the flash/save frequency of Energy Total in Tasmota

### DISCUSSION THREAD

**Discussion by kubedzero** [arendst/Tasmota#17894] [Q&A] [answered]
Hi there, I have a Sonoff S31 switch running Tasmota 12.3.1.5 (and 12.2.2 before I believe) connected to a space heater. I move it around the house, unplugging it from mains power and plugging it back in a few times a day. 

I have a script collecting the "Energy Total" value in kWh every 60 seconds, which I use for energy visualization. 

I have an issue where the Energy Total seems to "reset" back to a lower value than the last reported value. I imagine Tasmota is storing Energy Total in RAM and writing it to non-volatile memory every once in a while.

See picture for a visualization of Energy Total over a two day period.

![Tasmota Energy Total reading](https://user-images.githubusercontent.com/51148401/216874449-9425337f-a011-4321-bf65-d93d89905ee9.png)

Is the persistent save frequency of the "Energy Today/Yesterday/Total" documented anywhere, and is there any way to change it? I'd really love to avoid losing the Energy Total readings every time I unplug the device. I could tolerate say 60 minutes of loss, meaning I would ideally write the Energy Total to flash memory every hour of every day.

I see mention of the SaveData command https://tasmota.github.io/docs/Commands/#management but am not clear if "Energy Total" falls into the "changed parameters" category or if it is something else. A note in the manual to disambiguate this would really help.  

Thanks!

**Comment by kubedzero** [comment, marked as answer]
I found my answer in https://github.com/arendst/Tasmota/discussions/17825#discussioncomment-4815553 and https://github.com/arendst/Tasmota/issues/16791#issuecomment-1313584774

I can use Rules https://tasmota.github.io/docs/Commands/#rules to manually run the `savedata` command https://tasmota.github.io/docs/Commands/#management on a more frequent cadence 

`Rule1 on time#minute|60 do savedata endon` is an example Rule that apparently will save updated values every hour. I don't really get how that works, since `savedata 1` is apparently the default and as far as I can tell this is just calling the command with the default. Is there a side effect of calling this command that it works right away as well? Anyway, I am testing this now to confirm it works as expected.

```
22:32:34.546 CMD: Rule1 on time#minute|60 do savedata endon
22:32:34.551 RUL: Stored uncompressed, would compress from 35 to 23 (-35%)
22:32:34.554 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"on time#minute|60 do savedata endon"}}

22:33:20.960 CMD: Rule1 1
22:33:20.966 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"on time#minute|60 do savedata endon"}}
```

**Reply by kubedzero** [reply]
I just confirmed that unplugging the Sonoff at the 35 minute mark in an hour results in the loss of some of the Energy Total, but unplugging it within a minute of an hour (:00) results in no data loss. Furthermore, I checked the device's information page and confirmed that the "Flash Write Count" incremented by 1 within a few seconds of a new hour. 

This Rule works as expected. Thanks!

**Comment by arendst** [comment]
Yes. The side effect is that it requests drivers to save their changes to the settings RAM area and if there are any changes updates the flash accordingly.
